base/compiler/typeinfer.jl

```diff
@@ -8,7 +8,134 @@ function typeinf(interp::AbstractInterpreter, result::InferenceResult, cached::B
     return typeinf(interp, frame)
 end
 
+module Timings
+
+using Core.Compiler: -, +, length, push!, pop!, @inline
+
+struct Timing
+    name::Any
+    start_time::UInt64
+    cur_start_time::UInt64
+    time::UInt64
+    children::Core.Array{Timing,1}
+    Timing(name, start_time, cur_start_time, time, children) = new(name, start_time, cur_start_time, time, children)
+    Timing(name, start_time) = new(name, start_time, start_time, UInt64(0), Timing[])
+end
+
+time_ns() = ccall(:jl_hrtime, UInt64, ())
+
+const _timings = Timing[]
+function reset_timings()
+    Core.Compiler.empty!(_timings)
+    Core.Compiler.push!(_timings, Timing("root", time_ns()))
+    nothing
+end
+reset_timings()
+
+function close_current_timer()
+    stop_time = time_ns()
+    parent_timer = _timings[end]
```

Comment:
I guess this works because inference uses a depth-first search?

Reply:
Yeah, exactly, you're right.

I _believe_ that it's using a depth-first search, since everything seems to be implemented via recursive calls, rather than via an explicit queue. I'm pretty sure that's true. But you're totally right that this wouldn't work if it switched to some other traversal order. That's definitely worth calling out in comments; lemme do that now. :) thanks.